### PR TITLE
add a MultiStreamConn interface

### PR DESCRIPTION
**Pull request by marten-seemann**
Needed for the QUIC merge later on.

### CODE REVIEW

**Comment by marten-seemann** [comment]
This set of PRs should now be ready for review. Pinging @whyrusleeping, @Stebalien and @Kubuxu here.

This PR introduces a `SingleStreamConn` and a `MultiStreamConn` interface. TCP for example is a single stream connection, and QUIC is a multi-stream connection, since it supports native stream multiplexing.
All the connection setup (running the secio handshake and setting up the stream multiplexer, for single stream conns) happens in https://github.com/libp2p/go-libp2p-conn/pull/9. I also moved timeout-related code to that package as well.

Note that this PR does NOT add support for QUIC, due to the MITM issues discussed in https://github.com/libp2p/go-libp2p/issues/34. I will send a separate PR for that once QUIC supports TLS 1.3. After merging these set of PRs, we won't need any more interface changes in libp2p to support QUIC.

**Comment by ghost** [comment]
The MultiStreamConn has nothing to do with multiformats/multistream right?

**Comment by Stebalien** [comment]
@lgierth no, it doesn't. It's a conn interface that has a built-in stream muxer.

**Review by Stebalien — DISMISSED** [review]
LGTM

File: transport.go
Comment (Stebalien) [inline transport.go]: We can still assert that this implements the correct interfaces:

```go
var _ Conn = (MultiStreamConn)(nil)
```

File: transport.go
Comment (Stebalien) [inline transport.go:18]: Are we still planning on using this interface or deprecating it entierly? If we still want to ensure we meet this interface, it would probably be a good idea to assert that `SingleStreamConn` conforms to it.

**Review by Stebalien — COMMENTED** [review]
This mostly looks good to me I just want to make sure I understand what's going on here.

To use these connections, it looks like they need to be cast to either `SingleStreamConn`/`MultiStreamConn`. Is that correct?

Random comments (these aren't suggestions, just conversation starters; I'm not even sure I agree with myself here):

* Personally, I somewhat prefer the names `DuplexConn` and `MultiplexConn` but that's probably just personal preference and not worth bikeshedding over. 

* ~Have you considered making `MultiStreamConn` a superset of `SingleStreamConn`? You could do this by providing a "default" stream. The extra complexity may not be worth it but some consumers of this library may not care about or want to deal with streams (but may still want to use protocols like QUIC for other reasons). On the other hand, one could argue that users shouldn't be using transports directly (they should use go-libp2p-swarm instead).~ *Edit: I've discussed this with @whyrusleeping offline. Making MultiStreamConn inherit SingleStreamConn has been considered and rejected as it adds unwarranted complexity.*

I don't want to derail this PR but I'd like to at least air these questions first.

---

FYI, `test/utils.go` doesn't currently build.

File: transport.go
Comment (whyrusleeping) [inline transport.go]: maybe `MuxingConn` ?

Comment (marten-seemann) [reply transport.go]: I'd be fine with that, but then we probably should also rename the `SingleStreamConn` as well. Any suggestions for that?

Comment (whyrusleeping) [reply transport.go]: hrm... maybe `DirectConn` or `BaseConn`? @Stebalien ?

Comment (Stebalien) [reply transport.go]: To me, `BaseConn` implies that all other conns inherit from it and `DirectConn` implies that the connection is direct (which is true for all connections).

How about:

* `StreamingConn` (like `Muxing`)
* `BasicConn` (doesn't imply that it's the "base", just a minimal feature set).

(I still kind of like `DuplexConn`/`MultiplexConn` or `DuplexedConn`/`MultiplexedConn` but those are kind of long).

Comment (Stebalien) [reply transport.go]: @marten-seemann

So, I'd like to make progress on this set of PRs (both for QUIC and because they're holding up some other updates) but neither @whyrusleeping nor I really care about the names of these connections... we just don't like MultiStreamConn because we already have a concept of multistreams (and they're completely unrelated).

So... you decide (roll dice if you want). Just not MultiStreamConn.

Comment (marten-seemann) [reply transport.go]: I picked `DuplexConn` and `MultiplexConn`, and updated all PRs where those were used.

File: transport.go
Comment (marten-seemann) [reply transport.go]: Done.

**Comment by marten-seemann** [comment]
@Stebalien: Yes, unfortunately we will need a type assertion to actually use the connection. In principle, this could be avoided by doing the split into duplex and multi earlier, i.e. by introducing a `DuplexDialer` and a `MultiplexDialer` (and the corresponding listeners). I don't think this will simplify anything though, because we'd still have to do a type assertion somewhere.

**Review by Stebalien — APPROVED** [review]
LGTM

**Comment by Stebalien** [comment]
Thanks!

**Pull request merged**
